src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -13,81 +13,31 @@
 namespace stellar
 {
 
-std::vector<std::pair<Asset, Asset>>
-extractAssetPairs(Operation const& op, LedgerTxnDelta const& ltxd)
+static std::set<std::pair<Asset, Asset>>
+extractAssetPairs(LedgerTxnDelta const& ltxd)
 {
-    switch (op.body.type())
-    {
-    case MANAGE_BUY_OFFER:
-    {
-        auto const& offer = op.body.manageBuyOfferOp();
-        return {std::make_pair(offer.selling, offer.buying)};
-    }
-    case MANAGE_SELL_OFFER:
-    {
-        auto const& offer = op.body.manageSellOfferOp();
-        return {std::make_pair(offer.selling, offer.buying)};
-    }
-    case CREATE_PASSIVE_SELL_OFFER:
-    {
-        auto const& offer = op.body.createPassiveSellOfferOp();
-        return {std::make_pair(offer.selling, offer.buying)};
-    }
-    case PATH_PAYMENT:
-    {
-        auto const& pp = op.body.pathPaymentOp();
-
-        // if no path, only have a pair between send and dest
-        if (pp.path.size() == 0)
-        {
-            return {std::make_pair(pp.sendAsset, pp.destAsset)};
-        }
-
-        // For send, dest, {A, B} we get: {send, A}, {A, B}, {B, dest}
-        std::vector<std::pair<Asset, Asset>> assets;
-
-        // beginning: send -> A
-        assets.emplace_back(pp.sendAsset, pp.path.front());
-        for (int i = 1; i < pp.path.size(); ++i)
-        {
-            // middle: A -> B
-            assets.emplace_back(pp.path[i - 1], pp.path[i]);
-        }
-        // end: B -> dest
-        assets.emplace_back(pp.path.back(), pp.destAsset);
 
-        return assets;
-    }
-    case ALLOW_TRUST:
+    std::set<std::pair<Asset, Asset>> assets;
+    for (auto const& entry : ltxd.entry)
     {
-        auto const& at = op.body.allowTrustOp();
-
-        // if revoke auth, all offers for that user against that asset are
-        // deleted
-        if (!at.authorize)
+        if (entry.first.type() == OFFER)
         {
-            std::vector<std::pair<Asset, Asset>> assets;
-            // since we only get one side of the asset pair from the operation,
-            // we derive the rest of the information from the
-            // LedgerTxnDelta entries
-            for (auto const& entry : ltxd.entry)
+            auto const& offer = entry.second.previous
+                                    ? entry.second.previous->data.offer()
+                                    : entry.second.current->data.offer();
+
+            auto assetPair = std::make_pair(offer.selling, offer.buying);
+            auto oppositeAssetPair =
+                std::make_pair(offer.buying, offer.selling);
+            if (assets.find(oppositeAssetPair) == assets.end() &&
```

Comment:
Technically, you only need to check `oppositeAssetPair` here. The insert on line 35 will do nothing if `assetPair` is already included.

Reply:
Good point -- fixed.